### PR TITLE
Develop ttv

### PR DESCRIPTION
**Pull request by jvmafra**
Implementing TTV (Time to violate) preemption policy.

### CODE REVIEW

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+VM_BASE_HAS_GREATER_TTV = -1
+
+
+def _compare_vms(vm, vm_base):
+    """
+    Method that compares two requests considering slo, ttv and submit time.
+    """
+
+    vm_slo = float(vm.get_scheduler_hint('slo'))
+    vm_base_slo = float(vm_base.get_scheduler_hint('slo'))
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: I got it.

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+def compare_vms_to_reallocate(vm, vm_base):
+    vm_submit_time = dateutil.parser.parse(vm.get_scheduler_hint('submit_time'))
+    vm_base_submit_time = dateutil.parser.parse(vm_base.get_scheduler_hint('submit_time'))
+
+    compare_vms = _compare_vms(vm, vm_base)
+
+    if compare_vms == 0:
+        if vm_submit_time > vm_base_submit_time:
+            return vm_comparator.VM_BASE_ARRIVED_FIRST
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Where this constant was declared? I think a comment here would make things clearer. What does "VM_BASE_ARRIVED_FIRST" mean? Should VM_BASE be considered smaller?

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+        if vm_submit_time < vm_base_submit_time:
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Same here.

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+def compare_vms_to_preempt(vm, vm_base):
+    vm_submit_time = dateutil.parser.parse(vm.get_scheduler_hint('submit_time'))
+    vm_base_submit_time = dateutil.parser.parse(vm_base.get_scheduler_hint('submit_time'))
+
+    compare_vms = _compare_vms(vm, vm_base)
+
+    if compare_vms == 0:
+        if vm_submit_time > vm_base_submit_time:
+            return VM_BASE_ARRIVED_FIRST_PREEMPT
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Same here. I think a comment before the if clause would help to understand. Something like "If SLO and TTV are equal in both VMs, they will be sorte from XXXX (lower|greater) submit time to XXXX. It means that the vm with XXX will be preempted before than other."

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        return False
+
+    if slo_running_request < slo_new_request:
+        return True
+
+    current_time = datetime.datetime.now()
+
+    ttv_running_spec = calculate_ttv(vm_running, current_time)
+
+    if cant_wait(ttv_running_spec):
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: I think isViolatingSLO is a better name for this method

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        return False
+
+    ttv_new_spec = calculate_ttv(vm_request, current_time)
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: The idea is that ttv_new_spec must be a negative value to be possible to preempt vms with the same priority. I think that this verification should be in the bellow if clause.

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 5 above)
Comment (alessandroliafook) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Ok

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    runtime = float(vm.get_scheduler_hint('actual_runtime'))
+
+    submit_time = dateutil.parser.parse(vm.get_scheduler_hint('submit_time'))
+    status = vm.get_scheduler_hint('status')
+    last_start_exec = vm.get_scheduler_hint('last_start')
+    vm_slo = float(vm.get_scheduler_hint('slo'))
+
+    if status == allocation_status.RUNNING:
+        last_start = dateutil.parser.parse(last_start_exec)
+        print "Diff between last start and submit: %s" % str((last_start - submit_time).total_seconds())
```
Comment (alessandroliafook) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Removed!

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 2 above)
Comment (alessandroliafook) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Done

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 3 above)
Comment (alessandroliafook) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Done

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 4 above)
Comment (alessandroliafook) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Done

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        runtime += (current_time - last_start).total_seconds()
+
+    time_in_system = (current_time - submit_time).total_seconds()
+    print "Time in system: %s" % str(time_in_system)
+    print "Submit time = %s" % str(submit_time)
+    print "Current time = %s" % str(current_time)
+    print "Runtime = %s" % str(runtime)
```
Comment (alessandroliafook) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Done

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 6 above)
Comment (alessandroliafook) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Done

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    if last_start_exec == "":
```
Comment (jvmafra) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: No. last_start_exec == "" implies that the VM is arriving now, so its TTV is equals to INF. The same occur in availability aware policy. We are not dealing with VMs that aren't allocated initially, but if we are goint to treat it, we need to put a value in this attribute to identify this case, like -1, for example.

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 1 above)
Comment (alessandroliafook) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: OK

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    if vm_ttv > vm_base_ttv:
+        return vm_comparator.VM
+
+    if vm_ttv < vm_base_ttv:
+        return vm_comparator.VM_BASE
+
+    return vm_comparator.VMS_EQUALS
+
+
+def compare_vms_to_reallocate(vm, vm_base):
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Refactor both compare methods as we discussed.

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+    else:
+        ttv = runtime - (time_in_system * vm_slo)
+
+    print "The time to violate slo is: %.6f" % ttv
+
+    return ttv
+
+
+def preempt_vm(vm_request, vm_running):
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: This methods returns if vm_running is preemptable or not based on vm_request, correct? Would is_preemptable_vm(...) a better name for it?

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    Runtime = %s" % (str(time_in_system), str(submit_time), str(current_time), str(runtime))
+
+    if last_start_exec == "":
+        print "Last start is empty. New vm arriving"
+        ttv = float("inf")
+
+    else:
+        ttv = runtime - (time_in_system * vm_slo)
+
+    print "The time to violate slo is: %.6f" % ttv
```
Comment (giovannifs) [inline nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Join the prints of this method into only one.

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 11 above)
Comment (giovannifs) [reply nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: I think we should change this name in other policies as well.

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 12 above)
Comment (alessandroliafook) [reply nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Done

File: nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py (hunk 11 above)
Comment (alessandroliafook) [reply nova/scheduler/preemption_policy/time_to_violate_preemption_policy.py]: Done

**Pull request merged**